src/bin/rm.rs

```diff
@@ -1,21 +1,201 @@
+            }
+        }
+    }
 
-    if env::args().count() < 2 {
-        fail("no arguments.", &mut stderr);
+    /// Attempt to remove the file given as an input argument.
+    fn remove_file(&mut self, file: &Path) {
+        if let Err(message) = fs::remove_file(file) {
+            self.stderr.write(b"cannot remove '").try(&mut self.stderr);
+            self.stderr.write(file.to_string_lossy().as_bytes()).try(&mut self.stderr);
+            self.stderr.write(b"': ").try(&mut self.stderr);
+            print_error(message, &mut self.stderr);
+        } else if self.verbose {
+            self.stdout.write(b"removed '").try(&mut self.stderr);
+            self.stdout.write(file.to_string_lossy().as_bytes()).try(&mut self.stderr);
+            self.stdout.write(b"'\n").try(&mut self.stderr);
+            self.stdout.flush().try(&mut self.stderr);
+        }
     }
 
-    for ref path in env::args().skip(1) {
-        fs::remove_file(path).try(&mut stderr);
+    /// Attempt to remove a directory and all of it's contents if recursive mode is enabled.
+    /// If recursion is not enabled, attempt to remove the directory if it is empty.
+    fn remove_directory(&mut self, directory: &Path) {
+        // TODO: Use walkdir when it is implemented in Redox instead of fs::remove_dir_all().
+        if self.recursive {
+            if let Err(message) = fs::remove_dir_all(directory) {
+                self.stderr.write(b"cannot remove directory '").try(&mut self.stderr);
+                self.stderr.write(directory.to_string_lossy().as_bytes()).try(&mut self.stderr);
+                self.stderr.write(b"': ").try(&mut self.stderr);
+                print_error(message, &mut self.stderr);
+            } else if self.verbose {
+                self.stdout.write(b"removed directory '").try(&mut self.stderr);
+                self.stdout.write(directory.to_string_lossy().as_bytes()).try(&mut self.stderr);
+                self.stdout.write(b"'\n").try(&mut self.stderr);
+                self.stdout.flush().try(&mut self.stderr);
+            }
+        } else {
+            if let Err(message) = fs::remove_dir(directory) {
+                self.stderr.write(b"cannot remove directory '").try(&mut self.stderr);
+                self.stderr.write(directory.to_string_lossy().as_bytes()).try(&mut self.stderr);
+                self.stderr.write(b"': ").try(&mut self.stderr);
+                print_error(message, &mut self.stderr);
+            } else if self.verbose {
+                self.stdout.write(b"removed directory '").try(&mut self.stderr);
+                self.stdout.write(directory.to_string_lossy().as_bytes()).try(&mut self.stderr);
+                self.stdout.write(b"'\n").try(&mut self.stderr);
+                self.stdout.flush().try(&mut self.stderr);
+            }
+        }
     }
 }
+
+/// Print the message given by an io::Error to stderr.
+fn print_error(message: io::Error, stderr: &mut Stderr) {
```

Comment:
I can't think of any other sane approach to the behavior other than how BSD and GNU Core Utilities operates. If you are wanting to abort just because one file produces an error then you also shouldn't remove **any** files unless all files are 100% guaranteed to be removable without error. That's not very feasible.

Having files left over files from a rm command because one file or directory had an error would be a nightmare to track and script. Those using these commands would have to resort to creating a list of files and then executing rm once for each file versus accepting multiple arguments, thereby making multiple arguments completely useless to implement. As an example, instead of `rm *` they would have to write `for file in *; do rm $file; done` to ensure that the removal process doesn't abort midway.

There's absolutely nothing wrong with not aborting immediately, and it has no bearing on the usage for people writing scripts either. Just by parsing the output of the `rm -v` command will tell them if a certain key file that they want is removed successfully, without destroying rm's usefulness.